flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/Timestamp.java

```diff
@@ -0,0 +1,179 @@
+
+	public int getNanoOfMillisecond() {
+		return nanoOfMillisecond;
+	}
+
+	/**
+	 * Obtains an instance of {@code Timestamp} from a millisecond.
+	 *
+	 * <p>This returns a {@code Timestamp} with the specified millisecond.
+	 * The nanoOfMillisecond field will be set to zero.
+	 *
+	 * @param millisecond the number of milliseconds since January 1, 1970, 00:00:00 GMT
+	 *                    A negative number is the number of milliseconds before
+	 *                    January 1, 1970, 00:00:00 GMT
+	 * @return an instance of {@code Timestamp}
+	 */
+	public static Timestamp fromEpochMillis(long millisecond) {
+		return new Timestamp(millisecond, 0);
+	}
+
+	/**
+	 * Obtains an instance of {@code Timestamp} from a millisecond and a nanoOfMillisecond.
+	 *
+	 * <p>This returns a {@code Timestamp} with the specified millisecond and nanoOfMillisecond.
+	 *
+	 * @param millisecond the number of milliseconds since January 1, 1970, 00:00:00 GMT
+	 *                    A negative number is the number of milliseconds before
+	 *                    January 1, 1970, 00:00:00 GMT
+	 * @param nanoOfMillisecond the nanosecond within the millisecond, from 0 to 999,999
+	 * @return an instance of {@code Timestamp}
+	 */
+	public static Timestamp fromEpochMillis(long millisecond, int nanoOfMillisecond) {
+		return new Timestamp(millisecond, nanoOfMillisecond);
+	}
+
+	/**
+	 * Convert this {@code Timestmap} object to a {@link java.sql.Timestamp}.
+	 *
+	 * @return An instance of {@link java.sql.Timestamp}
+	 */
+	public java.sql.Timestamp toTimestamp() {
+		return java.sql.Timestamp.valueOf(toLocalDateTime());
+	}
+
+	/**
+	 * Obtains an instance of {@code Timestamp} from an instance of {@link java.sql.Timestamp}.
+	 *
+	 * <p>This returns a {@code Timestamp} with the specified {@link java.sql.Timestamp}.
+	 *
+	 * @param ts an instance of {@link java.sql.Timestamp}
+	 * @return an instance of {@code Timestamp}
+	 */
+	public static Timestamp fromTimestamp(java.sql.Timestamp ts) {
+		return fromLocalDateTime(ts.toLocalDateTime());
+	}
+
+	/**
+	 * Convert this {@code Timestamp} object to a {@link LocalDateTime}.
+	 *
+	 * @return An instance of {@link LocalDateTime}
+	 */
+	public LocalDateTime toLocalDateTime() {
+		int date = (int) (millisecond / MILLIS_PER_DAY);
+		int time = (int) (millisecond % MILLIS_PER_DAY);
+		if (time < 0) {
+			--date;
+			time += MILLIS_PER_DAY;
+		}
+		long nanoOfDay = time * 1_000_000L + nanoOfMillisecond;
+		LocalDate localDate = LocalDate.ofEpochDay(date);
+		LocalTime localTime = LocalTime.ofNanoOfDay(nanoOfDay);
+		return LocalDateTime.of(localDate, localTime);
+	}
+
+	/**
+	 * Obtains an instance of {@code Timestamp} from an instance of {@link LocalDateTime}.
+	 *
+	 * <p>This returns a {@code Timestamp} with the specified {@link LocalDateTime}.
+	 *
+	 * @param dateTime an instance of {@link LocalDateTime}
+	 * @return an instance of {@code Timestamp}
+	 */
+	public static Timestamp fromLocalDateTime(LocalDateTime dateTime) {
```

Comment:
Have you tested the "1969/01/05"? I am not sure the logical is right.

Reply:
Do you mean:
`
		LocalDateTime ldt = LocalDateTime.of(1969, 01, 05, 0, 0);
		Assert.assertEquals(ldt, Timestamp.fromLocalDateTime(ldt).toLocalDateTime());
		Assert.assertEquals("1969-01-05T00:00", Timestamp.fromLocalDateTime(ldt).toString());
`

Reply:
"1969/01/05T05:02...." I just want to test the negative value with nanos.

Reply:
Already tested for Timestamp/LocalDateTime before "1970-01-01 00:00:00"